### PR TITLE
Use custom event to trigger multus configuration

### PR DESCRIPTION
**Pull request by gatici**
# Description

This PR updates the multus library to include the fix for https://github.com/canonical/kubernetes-charm-libraries/issues/16 and implements the necessary changes in the charm part.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have bumped the version of the library

### CODE REVIEW

File: lib/charms/kubernetes_charm_libraries/v0/multus.py [hunk 1]
```diff
@@ -468,6 +502,7 @@ def __init__(
         container_name: str,
         cap_net_admin: bool = False,
         privileged: bool = False,
+        refresh_event: Optional[Union[BoundEvent, List[BoundEvent]]] = None,
```
Comment (gruyaume) [inline lib/charms/kubernetes_charm_libraries/v0/multus.py:505]: Why make it optional? My recommendation is to have it simply be of type BoundEvent. No need to support list of events at the moment, afaik there is no use case for this right now. This should make the rest of the init quite simple:

```python
self.framework.observe(refresh_event, self._configure_multus)
```

File: lib/charms/kubernetes_charm_libraries/v0/multus.py [hunk 2]
```diff
@@ -619,3 +662,7 @@ def _on_remove(self, event: RemoveEvent) -> None:
                 self.kubernetes.delete_network_attachment_definition(
                     name=network_attachment_definition.metadata.name  # type: ignore[union-attr]
                 )
+
+    def delete_pod(self) -> None:
+        """Delete the pod."""
+        self.kubernetes.delete_pod(self._pod)
```
Comment (gruyaume) [inline lib/charms/kubernetes_charm_libraries/v0/multus.py:668]: Should this change be part of this PR?

File: lib/charms/kubernetes_charm_libraries/v0/multus.py [hunk 3]
```diff
@@ -148,6 +160,15 @@ def __init__(self, namespace: str):
...
 
+    def delete_pod(self, pod_name: str) -> None:
+        """Deleting given pod.
+
+        Args:
+            pod_name (str): Pod name
+
+        """
+        self.client.delete(Pod, pod_name, namespace=self.namespace)
+
```
Comment (gruyaume) [inline lib/charms/kubernetes_charm_libraries/v0/multus.py:171]: Should this change be part of this PR?

File: lib/charms/kubernetes_charm_libraries/v0/multus.py [hunk 4]
```diff
@@ -457,6 +478,19 @@ def _container_security_context_is_set(
...
+    def __init__(self, handle: Handle):
+        super().__init__(handle)
+
+
+class NadConfigChangedCharmEvents(CharmEvents):
+    """NAD config changed events."""
+
+    nad_config_changed = EventSource(NadConfigChangedEvent)
+
+
```
Comment (gruyaume) [inline lib/charms/kubernetes_charm_libraries/v0/multus.py:493]: I'm not sure we need to have this at all, users should be able to pass any event (ex. ConfigChanged or a custom event) and our code should not care.

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: lib/charms/kubernetes_charm_libraries/v0/multus.py (hunk 2 above)
Comment (gatici) [reply lib/charms/kubernetes_charm_libraries/v0/multus.py:668]: This belongs to another PR, it should be removed.

File: lib/charms/kubernetes_charm_libraries/v0/multus.py (hunk 3 above)
Comment (gatici) [reply lib/charms/kubernetes_charm_libraries/v0/multus.py:171]: This belongs to another PR, it should be removed.

File: lib/charms/kubernetes_charm_libraries/v0/multus.py (hunk 4 above)
Comment (gatici) [reply lib/charms/kubernetes_charm_libraries/v0/multus.py:493]: There is a reason to define the a custom events in the library part. The event to be observed will be a custom event as native charm events does not appropriate for our situation. If we do not define any custom event in the library side, each consumer charm needs to define custom events to be passed and observed. This requires more modifications and repeated work in the several consumer charms. 
If a custom event exist in the library part which is created for this purpose, charms can import and use it easily.

File: lib/charms/kubernetes_charm_libraries/v0/multus.py (hunk 1 above)
Comment (gatici) [reply lib/charms/kubernetes_charm_libraries/v0/multus.py:505]: I tried to support the single Bound event or list of Bound events by considering the future requirements. At the moment, single event is enough. We can support list of events when we need.

File: lib/charms/kubernetes_charm_libraries/v0/multus.py (hunk 4 above)
Comment (gatici) [reply lib/charms/kubernetes_charm_libraries/v0/multus.py:493]: This is removed from the library side.

**Comment by gatici** [comment]
LXD version will be updated to 5.17/stable.

**Review by DanielArndt — CHANGES_REQUESTED** [review]

**Comment by DanielArndt** [comment]
Just the one small comment. Otherwise LGTM.

File: src/charm.py [hunk 5]
```diff
@@ -60,9 +60,24 @@ class IncompatibleCPUError(Exception):
     pass
 
 
+class NadConfigChangedEvent(EventBase):
+    """Event triggered when an existing network attachment definition is changed."""
+
+    def __init__(self, handle: Handle):
+        super().__init__(handle)
```
[the remark this answers is not in the document]
Comment (gatici) [reply src/charm.py]: Done

**Review by gruyaume — APPROVED** [review]

**Review by DanielArndt — APPROVED** [review]
Ahh I guess with a docstring you don't even need `pass`. TIL.

**Comment by gatici** [comment]
> Ahh I guess with a docstring you don't even need `pass`. TIL.

I will remove the init method of NadConfigChangedEvent from Docstring in the library part.

**Pull request merged**